neqo-qpack/src/table.rs

```diff
@@ -6,31 +6,46 @@
 
 use crate::static_table::{StaticTableEntry, HEADER_STATIC_TABLE};
 use crate::{Error, QPackSide, Res};
-use std::collections::{HashMap, VecDeque};
+use neqo_common::qtrace;
+use std::collections::VecDeque;
+use std::convert::TryFrom;
+
+pub struct LookupResult {
+    pub index: u64,
+    pub static_table: bool,
+    pub value_matches: bool,
+}
 
 #[derive(Debug)]
 pub struct DynamicTableEntry {
     base: u64,
     name: Vec<u8>,
     value: Vec<u8>,
-    refs: HashMap<u64, u8>, //TODO multiple header. value will be used for that: or of flags 0x1 for headers, ox2 for trailes.
+    /// Number of streams that refer this entry.
+    refs: u64,
 }
 
 impl DynamicTableEntry {
     pub fn can_reduce(&self, first_not_acked: u64) -> bool {
-        self.refs.is_empty() && self.base < first_not_acked
+        self.refs == 0 && self.base < first_not_acked
     }
 
     pub fn size(&self) -> u64 {
         (self.name.len() + self.value.len() + 32) as u64
     }
 
-    pub fn add_ref(&mut self, stream_id: u64, _block: u8) {
-        self.refs.insert(stream_id, 1);
+    pub fn add_ref(&mut self) {
+        self.refs += 1;
     }
 
-    pub fn remove_ref(&mut self, stream_id: u64, _block: u8) {
-        self.refs.remove(&stream_id);
+    pub fn remove_ref(&mut self) -> Res<()> {
+        if self.refs == 0 {
+            // This should never happen, this is a bug in our code.
+            debug_assert!(false);
+            return Err(Error::InternalError);
+        }
+        self.refs -= 1;
+        Ok(())
```

Comment:
I think long term `remove_ref` should not return Res. If we want to while this code is in development then I guess I'm ok but please add a qerror, and at some point I hope we can feel confident enough to stick an assert in here and not return Res.

Reply:
I will make the code to assert, let's find all the bus now, it is easier.